# [E-02] Classfifcation_breast_cancer  

# 미니 프로젝트 (3) load_breast_cancer : 유방암을 분류해 보자

## 첫 시도

cloud 환경이 아닌 local(m1 맥북) 환경에서 테이스 한 결과임을 참고 바란다.

sklearn breast_cancer dataset을 활용하여 분류해 보자.


### 1. 모듈 import

In [1]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

### 2. 데이터 준비  

- load_breast_cancer 메서드를 사용합니다.
- br_cancer_data에 Feature data 지정
- br_cancer_label에 Label(target) data 지정


In [2]:
br_cancer = load_breast_cancer()
br_cancer_data = br_cancer.data
print(br_cancer_data.shape, '\n')

br_cancer_label = br_cancer.target
print(br_cancer_label.shape, '\n')
# print(br_cancer_label[:20],'\n')
br_cancer.target

(569, 30) 

(569,) 



array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
       1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0,
       0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1,
       1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0,
       0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0,
       1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0,

- br_cancer_data에 569개의 데이터가 각각 30개의 정보를 담고 있다.
- br_cancer_label에도 569개의 데이터가 있음을 확인
- br_cancer.target에서 `0, 1` 2개의 class가 있음을 확인


### 3. 데이터 이해하기

다루어야 할 데이터를 자세히 살펴보자.

- Target Names 출력해 보기
- 데이터 Describe 해 보기


In [3]:
print(br_cancer.target_names)

['malignant' 'benign']


- breast_cancer 데이터 셋의 Target Names을 확인
- 2개의 class(악성, 양성)로 분류됨을 확인

In [4]:
print(br_cancer.DESCR)

.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 569

    :Number of Attributes: 30 numeric, predictive attributes and the class

    :Attribute Information:
        - radius (mean of distances from center to points on the perimeter)
        - texture (standard deviation of gray-scale values)
        - perimeter
        - area
        - smoothness (local variation in radius lengths)
        - compactness (perimeter^2 / area - 1.0)
        - concavity (severity of concave portions of the contour)
        - concave points (number of concave portions of the contour)
        - symmetry
        - fractal dimension ("coastline approximation" - 1)

        The mean, standard error, and "worst" or largest (mean of the three
        worst/largest values) of these features were computed for each image,
        resulting in 30 features.  For instance, field 0 is Mean Radi

- breast_cancer 데이터셋의 설명을 통해서도 기본적인 정보확인 가능
- 개의 class : 0(212개), 1(357개) 확인

### 4. 데이터 분리

In [5]:
X_train, X_test, y_train, y_test = train_test_split(br_cancer_data,
                                                   br_cancer_label,
                                                   test_size=0.2,
                                                   random_state=15)

print('X_train 개수: ', len(X_train),', X_test 개수: ', len(X_test), '\n')
print(X_train.shape, y_train.shape)
print(X_test.shape, y_test.shape)

X_train 개수:  455 , X_test 개수:  114 

(455, 30) (455,)
(114, 30) (114,)


- br_cancer 데이터 셋을 학습용 데이터와 테스트용 데이터로 분리
- test 데이터 셋의 크기는 전체 데이터셋의 20% 적용하여
- 학습용 데이터 455개, 테스트용 데이터 114개로 분리
- 데이터 분리할 때 랜덤으로 섞기 위해 random_state=15 적용

### 5. 다양한 모델로 학습시키기

1. Decision Tree 사용해 보기
2. Random Forest 사용해 보기
3. SVM 사용해 보기
4. SGD Classifier 사용해 보기
5. Logistic Regression 사용해 보기

In [6]:
# 각 모델 import
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn import svm
from sklearn.linear_model import SGDClassifier
from sklearn.linear_model import LogisticRegression

# 모델별 변수에 저장
decision_tree = DecisionTreeClassifier(random_state=15)
random_forest = RandomForestClassifier(random_state=15)
svm_model = svm.SVC()
sgd_model = SGDClassifier()
lr_model = LogisticRegression(max_iter=4000)  # digits작업시 warning message 해결방법 적용

### 5.1 Decision Tree

In [7]:
decision_tree.fit(X_train, y_train)
y_pred_decision_tree = decision_tree.predict(X_test)

print(classification_report(y_test, y_pred_decision_tree))

              precision    recall  f1-score   support

           0       0.97      0.87      0.92        39
           1       0.94      0.99      0.96        75

    accuracy                           0.95       114
   macro avg       0.95      0.93      0.94       114
weighted avg       0.95      0.95      0.95       114



### Decision Tree를 이용한 결과

- accuracy : 95%


### 5.2 Random Forest

In [8]:
random_forest.fit(X_train, y_train)
y_pred_random_forest = random_forest.predict(X_test)

print(classification_report(y_test, y_pred_random_forest))

              precision    recall  f1-score   support

           0       0.94      0.85      0.89        39
           1       0.92      0.97      0.95        75

    accuracy                           0.93       114
   macro avg       0.93      0.91      0.92       114
weighted avg       0.93      0.93      0.93       114



### Random Forest를 이용한 결과

- accuracy : 93%


### 5.3 SVM

In [9]:
svm_model.fit(X_train, y_train)
y_pred_svm_model = svm_model.predict(X_test)

print(classification_report(y_test, y_pred_svm_model))

              precision    recall  f1-score   support

           0       0.93      0.67      0.78        39
           1       0.85      0.97      0.91        75

    accuracy                           0.87       114
   macro avg       0.89      0.82      0.84       114
weighted avg       0.88      0.87      0.86       114



# SVM을 이용한 결과

- accuracy : 87%


### 5.4 SGD Classifier

In [10]:
sgd_model.fit(X_train, y_train)
y_pred_sgd_model = sgd_model.predict(X_test)

print(classification_report(y_test, y_pred_sgd_model))

              precision    recall  f1-score   support

           0       0.88      0.74      0.81        39
           1       0.88      0.95      0.91        75

    accuracy                           0.88       114
   macro avg       0.88      0.85      0.86       114
weighted avg       0.88      0.88      0.87       114



### SGD Classifier를 이용한 결과

- accuracy : 85%


### 5.5 Logistic Regression

In [11]:
lr_model.fit(X_train, y_train)
y_pred_lr_model = lr_model.predict(X_test)

print(classification_report(y_test, y_pred_lr_model))

              precision    recall  f1-score   support

           0       0.89      0.82      0.85        39
           1       0.91      0.95      0.93        75

    accuracy                           0.90       114
   macro avg       0.90      0.88      0.89       114
weighted avg       0.90      0.90      0.90       114



### Logistic Regression을 이용한 결과

- accuracy : 90%

### 결론

- #### accuracy  


    - Decision Tree : 95%
    - Random Forest : 93%
    - SVM : 87%
    - SGD Classifier : 85%
    - Logistic Regression : 90% 

- 모델의 성능을 평가하는 척도로 위와 같은 유방암의 경우에는 단순하게 **정확도(accuracy)**를 이용하는 것은 적합하지 않은 방법으로 판단된다.`암이 아닌데 암으로 판단하는 것`은 재검사를 통해서 밝힐 수도 있고 `실제 병이 있는 것이 아니기에` 생명에 지장이 있는 것은 아니지만, `암인데 암이 아닌것으로 판단하는 것은 생존의 문제와 직결되는 크나큰 문제`이므로 **confusion matrix의 recall(재현율)**을 선택하는 것이 적합한 것으로 판단된다.
- #### 각 모델 `0`(악성종양)의 recall(재현율) 현황


    - Decision Tree : 87%
    - Random Forest : 85%
    - SVM : 67%
    - SGD Classifier : 90%
    - Logistic Regression : 82%  


- SGD Classifier 모델의 Recall(재현율)이 90% 로 가장 높게 나왔다. 그러므로 **SGD Classifier**를 이용하여 유방암 여부를 예측하는 것이 가장 적합한 모델로 판단된다.